### PR TITLE
Categorical split support for SHAP

### PR DESCRIPTION
**Pull request by trivialfis**
Depends on https://github.com/rapidsai/gputreeshap/pull/26 .

### CODE REVIEW

File: src/predictor/gpu_predictor.cu [hunk 1]
```diff
@@ -481,20 +513,45 @@ void ExtractPaths(dh::device_vector<gpu_treeshap::PathElement>* paths,
...
+        if (is_left_path) {
+          for (size_t i = 0; i < node_cats.size(); ++i) {
+            path_cats[i] = ~node_cats[i];
+          }
+        } else {
+          for (size_t i = 0; i < node_cats.size(); ++i) {
+            path_cats[i] = node_cats[i];
+          }
+        }
+        // fixme: distinguish whether it's left path.
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu:539]: Remove this.

File: src/predictor/gpu_predictor.cu [hunk 2]
```diff
@@ -462,14 +465,43 @@ void ExtractPaths(dh::device_vector<gpu_treeshap::PathElement>* paths,
...
+                       return ft == FeatureType::kCategorical;
+                     })) {
+    auto max_elem_it = dh::MakeTransformIterator<size_t>(
+        dh::tbegin(d_cat_node_segments),
+        [] __device__(RegTree::Segment seg) { return seg.size; });
+    size_t max_cat_it =
+        thrust::max_element(thrust::device, max_elem_it,
+                            max_elem_it + d_cat_node_segments.size()) -
+        max_elem_it;
+    max_cat =
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu:490]: Use pinned memory.

File: src/predictor/gpu_predictor.cu [hunk 3]
```diff
@@ -481,20 +513,45 @@ void ExtractPaths(dh::device_vector<gpu_treeshap::PathElement>* paths,
...
-          upper_bound, is_missing_path,     zero_fraction, v};
+
+      float lower_bound = -inf;
+      float upper_bound = inf;
+      gpu_treeshap::Bitset bits;
+      if (common::IsCat(tree.cats.split_type, child.Parent())) {
+        auto path_cats = d_path_categories.subspan(max_cat * output_position, max_cat);
+        size_t size = tree.cats.node_ptr[child.Parent()].size;
+        auto node_cats = tree.cats.categories.subspan(tree.cats.node_ptr[child.Parent()].beg, size);
+        assert(path_cats.size() >= node_cats.size());
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu:529]: span check.

**Comment by trivialfis** [comment]
Please do not merge this one until I can replace the submodule back to rapidsai.

File: src/predictor/gpu_predictor.cu [hunk 4]
```diff
@@ -481,20 +575,44 @@ void ExtractPaths(dh::device_vector<gpu_treeshap::PathElement>* paths,
...
+      float upper_bound = inf;
+      common::CatBitField bits;
+      if (common::IsCat(tree.cats.split_type, child.Parent())) {
+        auto path_cats = d_path_categories.subspan(max_cat * output_position, max_cat);
+        size_t size = tree.cats.node_ptr[child.Parent()].size;
+        auto node_cats = tree.cats.categories.subspan(tree.cats.node_ptr[child.Parent()].beg, size);
+        SPAN_CHECK(path_cats.size() >= node_cats.size());
+        if (is_left_path) {
+          for (size_t i = 0; i < node_cats.size(); ++i) {
+            path_cats[i] = ~node_cats[i];
```
Comment (RAMitchell) [inline src/predictor/gpu_predictor.cu:594]: `path_cats[i] = is_left_path ? ~node_cats[i] : node_cats[i];`?

**Review by RAMitchell — APPROVED** [review]

File: src/predictor/gpu_predictor.cu (hunk 4 above)
Comment (trivialfis) [reply src/predictor/gpu_predictor.cu:594]: So in the bitset and xgboost predict, values that are set to 1 go to right, 0 go to left, and each bit corresponds to a category.   But when we construct the shap path, there's no left or right in the split condition, only "whether should we go down this path".  As a result, for paths that go to right node, we keep the bits unchanged, but for the paths that go to the left, we flip the bits.

Comment (trivialfis) [reply src/predictor/gpu_predictor.cu]: Done.

**Comment by talieh-tabatabaei** [comment]
How Shap values for each one-hot-encoded category are aggregated to generated  the final value for the categorical variables?

**Comment by oOTWK** [comment]
@talieh-tabatabaei When `enable_categorical=True` is enabled, it does not compute Shap values for each one-hot-encoded variable. To enable this for Shap computation, the tree model should be trained with this enabled. In this case, there are no one-hot-encoded variables anymore, it's just one categorical variable itself. So decision variable in the tree model is the categorical variable.

For example, let's say we have Fruit category and the values are Apple, Orange, and Melon. A node of tree could have Fruit==Orange. In one-hot-encoded version, it is equivalent to Is_Orange==1 but in this case, Is_Orange, Is_Apple, and Is_Melon are different variables. In TreeShap, these variables do not invoke `Unwind` to each other. On the other hand, if we just use the categorical variable Fruit, Fruit==Orange and Fruit==Apple and Fruit==Melon all use the same variable Fruit. This does invoke `Unwind` if there is more than one Fruit variable appearing on the current path. As a result, it will only compute Shap value for Fruit variable.

**Pull request merged**